### PR TITLE
Allow to extend IdleStateHandler and so provide more details for Idle…

### PR DESCRIPTION
**Pull request by normanmaurer**
…StateEvents

Motivation:

Sometimes it is useful to include more details in the IdleStateEvents that are produced by the IdleStateHandler. For this users should be able to create their own IdleStateEvents that encapsulate more informations.

Modifications:

- Make IdleStateEvent constructor protected and the class non-final
- Add protected method to IdleStateHandler that users can override and so create their own IdleStateEvents.

Result:

More flexible and customizable IdleStateEvents / IdleStateHandler

### CODE REVIEW

**Comment by normanmaurer** [comment]
@trustin PTAL ... @jchambers just for you ;)

**Comment by jchambers** [comment]
:metal:

Thank you!

**Comment by Scottmitch** [comment]
lgtm

File: handler/src/main/java/io/netty/handler/timeout/IdleStateEvent.java [hunk 1]
```diff
@@ -31,7 +31,13 @@
...
     private final boolean first;
 
-    private IdleStateEvent(IdleState state, boolean first) {
+    /**
+     * Constructor for sub-classes.
+     *
+     * @param state the {@link IdleStateEvent} which triggered the event.
+     * @param first {@code true} if its the first idle event for the {@link IdleStateEvent}.
+     */
+    protected IdleStateEvent(IdleState state, boolean first) {
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/timeout/IdleStateEvent.java]: @Scottmitch let me add some null checks tho...

**Comment by trustin** [comment]
LGTM

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (89eae94542ae3d522a78e1f8b0172782c9786901) and 4.0 (219bc64)

**Pull request closed without merging**